etc/nb-completion.fish

```diff
@@ -0,0 +1,169 @@
+
+        if eval set -q __desc_$__command
+          continue
+        end
+
+        # Read the description up to 40 chars.
+        set ___desc (echo $__help | string match -r 'Description:\t  (.*?)\.')[2]
+        if test -z $___desc
+          continue
+        end
+        set ___desc (string replace -a \t '' $___desc | string sub -l 40)
+        if test (string length $___desc) = 40
+          set ___desc (string sub -l 38 $___desc)……
+        end
+
+        eval set __desc_$__command \$___desc
+
+        # When __command is an alias for another command, also set the
+        # description for that.
+        set ___original_name (string match -r 'Usage:\s+nb\s+(\S+)' $___help)[2]
+        if test -n $___original_name
+          and test "$___original_name" != $__command
+          and not set -q __desc_$___original_name
+          eval set __desc_$___original_name \$___desc
+        end
+
+        # When __command has aliases, also set the description for them.
+        set ___aliases (string match -r 'Aliases: `([^\t]*)`' $___help)[2]
+        if test -n $___aliases
+          set ___aliases (string split '`, `' $___aliases)
+          for ___alias in $___aliases
+            if not eval set -q __desc_$___alias
+              eval set __desc_$___alias \$___desc
+            end
+          end
+        end
+      end
+
+      for __command in $_commands
+        if eval set -q __desc_$__command
+          eval set -a _completions \$__command\\t\$__desc_$__command
+        else
+          set -a _completions $_commands
+        end
+      end
+
+      # Construct <nootbook>:<subcommand> completions.
+      for __notebook in $_notebooks
+        for __command in $_commands
+          if test -n $__notebook
+            and test -n $__command
+            if eval set -q __desc_$__command
+              eval set -a _completions \$__notebook:\$__command\\t\$__desc_$__command
+            else
+              set -a _completions $__notebook:$__command
+            end
+          end
+        end
+      end
+
+      set _directory_path (dirname $_cache_path)
+      mkdir -p $_directory_path
+
+      begin
+        echo $_commands
+        echo $_notebooks
+        printf "%s\n" $_completions
+      end > $_cache_path
+    end
+
+    if set -q _flag_erase
+      rm $_flag_erase
+    end
+  end
+
+  set _nb_dir (nb env | string replace -fr "^NB_DIR=" "")
+
+  if test -z $_nb_dir
+    or not test -e $_nb_dir
+    return
+  else if test -L $_nb_dir
+    set _nb_dir (realpath $_nb_dir)
+  end
+
+  if not test -d $_nb_dir
+    return
+  end
+
+  set _cache_path $_nb_dir/.cache/nb-completion-cache-fish
+
+  if not test -e $_cache_path
+    _cache_completions $_cache_path
+  end
+
+  # list all notes to complete
+  nb run ls | string replace -r '$' \tfile
+
+  if test -e $_cache_path
+    tail -n+3 $_cache_path
+
+    # write the func itself into a temporary file and execute it in background
+    set _tmp_file (mktemp -t nb-completion)
```

Comment:
This throws errors "Too few Xs" . Change to "nb-completion.XXX"

Reply:
I found `mktemp` in Linux needs `X`s that one in macOS dose not need. Fixed on 30207a4.